### PR TITLE
proto2gds bin takes protos and makes them gds

### PR DESCRIPTION
**Pull request by growly**
Adds `proto2gds` binary, the inverse of `gds2proto`.

Reorganises `gds2proto` and `proto2gds` into `layout21converters` package instead of the `gds2proto` package.

### CODE REVIEW

**Review by dan-fritchman — CHANGES_REQUESTED** [review]
Can we also pull `lef2yaml` into this new `layout21converters`?

File: layout21converters/src/proto2gds.rs [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    };
+
+    if options.verbose {
+        println!("read: {:?}", &options.tech);
+    }
+
+    let mut layers_by_number = HashMap::new();
+
+    // Generate the Layers data from the given tech proto:
+    for layer_pb in tech_library.layers {
```
[the remark this answers is not in the document]
Comment (growly) [reply layout21converters/src/proto2gds.rs:62]: lef2yaml is now a bin in the layout21converters package.

As for where to move that conversion code, where would you like to see it live? To my mind it makes sense to put it in the [`from_proto`](https://github.com/dan-fritchman/Layout21/blob/404abfb1bf9a2e43f831c566603974939925ba71/layout21raw/src/proto.rs#L32) method: Another consistent organisation might be to give `struct Layers` its own `from_proto` method.

**Comment by growly** [comment]
ok, done. Not sure how to test this since the golden-file-byte-for-byte-comparison fails. GDS isn't exported deterministically? (I recall you had this problem once...)

**Comment by dan-fritchman** [comment]
I suspect the reason the GDSes don't match the golden version is this here: 
https://github.com/dan-fritchman/Layout21/blob/404abfb1bf9a2e43f831c566603974939925ba71/gds21/src/lib.rs#L809

In short: 
* The GDSII format has built-in creation and modification dates/ times. 
  * And! Their own [encoding of them](https://github.com/dan-fritchman/Layout21/blob/404abfb1bf9a2e43f831c566603974939925ba71/gds21/src/write.rs#L502).
  * Yeah it's old, and I guess one couldn't rely on the operating system for such things? 
* Absent a specified date/ time, Gds21 sets them to the current date & time by default. 

FWIW `gds21::GdsLibrary` does implement the `Eq` trait, so we can just compare them with things like `gdslib1 == gdslib2` and `assert_eq!(gdslib1, gdslib2)`. 

But it doesn't seem those dates/ times would ever match, for a utility like `proto2gds`.  
Couple things we could do: 

* Somehow hack the dates/ times for the test. Not my favorite. 
* Edit the `Eq` (really `PartialEq`) implementation to ignore the dates/ times. 
* Write another method that compares everything in the `GdsLibrary`, *except for* those dates/ times. 

I think we probably want the latter; checking whether two `GdsLibrary` are "fully identical" - including those dates - seems like something that will retain some utility.

**Comment by growly** [comment]
Have implemented `GdsLibrary::set_all_dates` and re-instated the golden file comparison; now it works again 👍

**Comment by dan-fritchman** [comment]
Yeah this is great. 
And after some more thought, we're doing the right thing _not_ making a comparison function that ignores the dates. 

Because, well, `GdsLibrary`s could be "equivalent" in all kinds of other ways, such as: 

* The list of `structs` (cells) could be in different orders
* The list of `elements` in any given `GdsStruct` could be in different orders
* Shapes could be replaced with any number of isomorphic equivalent things. 
  * E.g. replacing `Box` (a square) with a square-shaped `Polygon` or `Path`
  * E.g. rotating a square by any multiple of 90 degrees 
* Arrays could be flattened, or a flat list of instances could be arrayed

Any or all of these differences would represent the same layout.  
Actually deciding whether two GDSII files represent the same layout requires resolving all of them, and probably more that I didn't think of in these few minutes. 

So - I like `gds21::GdsLibrary`'s version of equality, which is essentially "would these two `GdsLibrary`s produce identical bits on disk". And I like the having adding the method "set all the timestamps" method.

**Comment by growly** [comment]
Agreed. Anything else you see needing changes here or is it go to merge?

**Comment by dan-fritchman** [comment]
Terrific, merge time. 

FWIW the "bootstrap the VLSIR dependency" did require some finicking with Python environments, and just what version of their packages are installed. It's a dev-mode only thing, hopefully we don't need it for long.

**Pull request merged**